src/Core/Batch/BatchTrait.php

```diff
@@ -0,0 +1,195 @@
+                    memory_get_usage()
+                )
+            );
+        }
+
+        return true;
+    }
+
+    /**
+     * Returns an array representation of a callback which will be used to write
+     * batch items.
+     *
+     * @return array
+     */
+    protected abstract function getCallback();
+
+    /**
+     * @param array $options [optional] {
+     *     Configuration options.
+     *
+     *     @type resource $debugOutputResource A resource to output debug output
+     *           to.
+     *     @type bool $debugOutput Whether or not to output debug information.
+     *           **Defaults to** false.
+     *     @type array $batchOptions A set of options for a BatchJob.
+     *           {@see \Google\Cloud\Core\Batch\BatchJob::__construct()} for
+     *           more details.
+     *           **Defaults to** ['batchSize' => 1000,
+     *                            'callPeriod' => 2.0,
+     *                            'workerNum' => 2].
+     *     @type array $clientConfig A config used to construct the client upon
+     *           which requests will be made.
+     *     @type BatchRunner $batchRunner A BatchRunner object. Mainly used for
+     *           the tests to inject a mock. **Defaults to** a newly created
+     *           BatchRunner.
+     *     @type string $identifier An identifier for the batch job.
+     *     @type string $batchMethod The name of the batch method used to
+     *           deliver items.
+     * }
+     * @throws \InvalidArgumentException
+     */
+    private function setCommonBatchProperties(array $options)
+    {
+        if (!isset($options['identifier'])) {
+            throw new \InvalidArgumentException(
+                'A valid identifier is required in order to register a job.'
+            );
+        }
+
+        if (!isset($options['batchMethod'])) {
+            throw new \InvalidArgumentException(
+                'A batchMethod is required.'
+            );
+        }
+
+        $this->batchMethod = $options['batchMethod'];
+        $this->identifier = $options['identifier'];
+        $this->debugOutputResource = isset($options['debugOutputResource'])
+            ? $options['debugOutputResource']
+            : null;
+        $this->debugOutput = isset($options['debugOutput'])
+            ? $options['debugOutput']
+            : false;
+        $this->clientConfig = isset($options['clientConfig'])
+            ? $options['clientConfig']
+            : [];
+        $batchOptions = isset($options['batchOptions'])
+            ? $options['batchOptions']
+            : [];
+        $this->batchOptions = $batchOptions + [
+            'batchSize' => 1000,
+            'callPeriod' => 2.0,
+            'workerNum' => 2
+        ];
+        $this->batchRunner = isset($options['batchRunner'])
+            ? $options['batchRunner']
+            : new BatchRunner();
+        $this->batchRunner->registerJob(
+            $this->identifier,
+            [$this, 'send'],
```

Comment:
Ok, I tested it, removed the `send` and `getCallback`, and directly registered the function. Unfortunately I couldn't achieve any significant performance gain.

Probably we can move forward with the current implementation and I will invest some of my time on the performance improvement in the future. Also, the end2end testing that I'm conducting is actually easily affected by the Logging API's server status, so the method is not very scientific.

The minimum bound of the performance we're getting with the current code is acceptable, so maybe we can move forward?